### ISSUE TITLE
Rule proposal: No unreadable array destructuring

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#197]
Destructuring is very useful, but it can also make some code harder to read.

For example: https://github.com/sindresorhus/refined-github/pull/1558/files#diff-7fa8bf1c42c5640497119c24113948a4R56

## Fail

```js
const [,, foo] = parts;
const [,,, foo] = parts;
const [,,,, foo] = parts;
// ...and so on...
```

## Pass

```js
const [, foo] = parts;
const [foo] = parts;
const foo = parts[3];
```

**Comment by MrHen** [comment]
Why is one comma okay? If the rule just prevented empty segments it would be pretty simple. Is the `[, foo]` case common enough to explicitly include?

Also, what about:

```
const [foo,,bar] = parts;
const [foo,,,,bar] = parts;
```

**Comment by sindresorhus** [comment]
> Why is one comma okay?

It felt readable enough, but I guess we could disallow that one too.

> Also, what about:

Yes, those too.

**Comment by voxpelli** [comment]
I think a comma in the beginning is reasonable to allow

**Comment by pvdlg** [comment]
Isn't this rule would conflict with [`prefer-destructuring`](https://eslint.org/docs/rules/prefer-destructuring)?
```js
const [,, foo] = parts;
```
Would trigger that rule.
```js
const foo = parts[2];
```
would trigger `prefer-destructuring`. So you would end up in a situation in which both solution are invalid as you probably don't want to disable `prefer-destructuring` which is useful in many other cases.

So should this rule be a re-implementation of `prefer-destructuring` that would work the same but with the exception or not erroring out for unreadable cases mentioned above?

Also should cases like `const [foo, , bar] = parts;` be valid? Maybe a better implementation would be to disallow multiple "missing" values in a row.
This would be an error:
```js
const [,, foo] = parts;
const [foo, , , bar] = parts;
```

This would be OK:
```js
const [, foo] = parts;
const [foo, , bar] = parts;
```

**Comment by MrHen** [comment]
`prefer-destructuring` has a setting to disable `array` destructuring which is almost certainly what you would want if you don't like `const [,, foo] = parts;`. From their docs:

> Additionally, if you intend to access large array indices directly, like:
>
> ```var foo = array[100];```
>
> Then the `array` part of this rule is not recommended, as destructuring does not match this use case very well.

If you want both rules enabled you would have to use `eslint-disable-next-line` to signal which rule you prefer for each instance.
